### ISSUE TITLE
--docker-registry doesn't work

### ISSUE THREAD

**Issue by evanchaoli** [concourse/concourse#3360]
## Bug Report

I'm trying to use our own docker artifactory as the default place to pull images. I did some try today on a 5.0 cluster.

By passing some environment variables to `concourse worker`, I can see that `gdn` has got `--docker-registry` option:

> root        26     6  0 07:45 ?        00:00:02 gdn server --bind-ip 0.0.0.0 --bind-port 7777 --depot /worker-state/depot --properties-path /worker-state/garden-properties.json --time-format rfc3339 --no-image-plugin --docker-registry my.artifactory.eng.vmware.com --insecure-docker-registry my.artifactory.eng.vmware.com --dns-server 172.18.0.4 --allow-host-access

Then I submit a basic pipeline to the cluster with specifying an image path of my docker artifactory, but Concourse failed to pull the image, so looks like `--docker-registry` doesn't work.

## Steps to Reproduce

Exports envs:

      CONCOURSE_GARDEN_DOCKER_REGISTRY=my.artifactory.eng.vmware.com
      CONCOURSE_GARDEN_INSECURE_DOCKER_REGISTRY=my.artifactory.eng.vmware.com

then start worker.

## Expected Results

Worker is expected to use my.artifactory.eng.vmware.com as the default registry.

## Actual Results

Cannot pull image from my.artifactory.eng.vmware.com.

## Version Info

* Concourse version: 5.0
* Deployment type (BOSH/Docker/binary): Docker
* Infrastructure/IaaS: docker-compose
* Browser (if applicable): any
* Did this used to work? no

**Comment by vito** [comment]
I'm not sure what you mean by a default registry, but this is not supported - it's a feature of Garden-runC that we intentionally don't expose, and in 5.0 we no longer show the flags (though you can still set it via env vars, as we allow all of them to be configured for backwards-compatibility).

You should probably be using [`image_resource`](https://concourse-ci.org/tasks.html#image_resource) instead.

**Comment by vito** [comment]
@isaguimiot This flag is for an internal component's code path that Concourse never exercises. It was never supposed to be exposed, and is now hidden.

The Docker Image resource supports configuring `registry_mirror`: https://github.com/concourse/docker-image-resource

**Comment by isaguimiot** [comment]
Hi @vito , thanks for your quick response.  It's a blocker issue for us, so I was hoping you could answer my two questions so we can analyze how to prevent our Concourse IPs from being blocked by DockerHub again: 
Could you tell us the reasons why this feature isn't available ? 
And what are the alternatives for us (I mean real alternatives, other than having over 1000 devs add a parameter in every docker-image resource ) ? 

Thank you !

**Comment by isaguimiot** [comment]
Hi !

We have a similar issue here : I configured garden runc in the concourse deployment, setting the "docker_registry_endpoint" parameter to point to our Artifactory registry mirror of docker hub, and when I run a pipeline in concourse, it still pulls it directly from docker hub :(

It works fine when I set the mirror directly into the image_resource part of the pipeline.yml, but we don't want all the users to modify their pipelines, we need to set it globally.

The point is that we were blacklisted by docker hub 2 months ago, because of the huge amount of pulls we were doing from concourse... A local mirror is the only way to prevent us from being banned again.

Could you tell us the reasons why this feature isn't available ? And what are the alternatives for us ?

Thank you !

**Comment by ukabu** [comment]
@vito we really need a way to configure Concourse to pull all images from a private docker registry without having to specify that in every pipelines and tasks.
We run more that 13000 jobs per weeks. Have about 150K resource checks per hour. More that 150 teams.
We're getting rate limited by DockerHub and we could easily solve this problem by having a private mirror.
This is becoming a major pain point for us.

**Comment by vito** [comment]
@ukabu Could you please open a new issue describing that use case?

**Comment by billimek** [comment]
@ukabu when you do, please reference it here too.  We are facing a similar issue.

**Comment by skreddy6673** [comment]
@ukabu Did you open any new issue for this use case or found any workarounds?

**Comment by ukabu** [comment]
We are modifying the bundled resource to point to our artifactory instance before deploying Concourse. This is a bit of a hack, but it works well.

**Comment by skreddy6673** [comment]
@ukabu Thank you for your reply. can i look at your bundle if possible?

**Comment by skreddy6673** [comment]
@vito Is this configuration available in any of the latest versions of concourse? The reason for ask is with new docker policies changing we are planning to use atifactory to be a Proxy of dockerhub.

**Comment by vito** [comment]
@skreddy6673 No - as I mentioned in https://github.com/concourse/concourse/discussions/6039#discussioncomment-61205 this will require something like the feature proposed in https://github.com/concourse/concourse/pull/5942#issuecomment-669283962.

I'll start an RFC for it - we need to get the ball rolling since the Docker Hub deadline is approaching.

**Comment by evanchaoli** [comment]
I will restart #5942 soon.

**Comment by vito** [comment]
@evanchaoli Thanks! I'm going to work on some short-term mitigations too. My plan is to make it so that image `check`s are given the latest version as the cursor and update `registry-image` and `docker-image` to just return the cursor version when they hit the rate limit. This should at least prevent builds from grinding to a halt, but it will only help for images which have already been `check`ed at least once.
